### PR TITLE
Clone Page feature inside an application

### PR DESCRIPTION
**Pull request by trishaanand**

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java [hunk 1]
```diff
@@ -324,4 +335,86 @@ private void generateAndSetPagePolicies(Application application, User user, Page
...
+                    Mono<ApplicationPagesDTO> pageNamesMono = pageService
+                            .findNamesByApplicationId(page.getApplicationId());
+                    return pageNamesMono
+                            // Set a unique name for the cloned page and then create the page.
+                            .flatMap(pageNames -> {
+                                Set<String> names = pageNames.getPages()
+                                        .stream()
+                                        .map(pageNameIdDTO -> pageNameIdDTO.getName()).collect(Collectors.toSet());
+
+                                String newPageName = page.getName() + "'s Copy";
```
Comment (mohanarpit) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java:372]:
```suggestion
                                String newPageName = page.getName() + " Copy";
```

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java [hunk 2]
```diff
@@ -324,4 +335,86 @@ private void generateAndSetPagePolicies(Application application, User user, Page
...
+                })
+                .flatMap(page -> {
+                    String newPageId = page.getId();
+                    return sourceActionFlux
+                            .flatMap(action -> {
+                                action.setId(null);
+                                action.setPageId(newPageId);
+                                return actionService.create(action);
+                            })
+                            .retry()
```
Comment (mohanarpit) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java:394]: Please remove this. Doesn't seem relevant.

**Review by mohanarpit — APPROVED** [review]

**Comment by sharat87** [comment]
Hey @trishaanand, sorry just got around to this. It looks like we are calling `LayoutActionService.updateLayout` so that the actions inside onLoad lists are updated to point to the new actions. However, this doesn't update the actions inside `publishedLayoutOnLoadActions`. So, action IDs inside `publishedLayoutOnLoadActions` would still be pointing to actions in the source page. They have to be manually mapped and updated.

Otherwise looks good to have been merged. Ignore if this is already handled.

**Comment by trishaanand** [comment]
Hey @sharat87 : The new page created would be in unpublished zone. So none of the published fields would be set including the published on load actions and published dsl.

**Comment by mohanarpit** [comment]
@all-contributors Please add @trishaanand  and @sharat87  for their contributions to code, infrastructure and ideas.

**Pull request merged**